drivers/platform/stm32/stm32_gpio_irq.c

```diff
@@ -0,0 +1,497 @@
+/***************************** Static variables *******************************/
+/******************************************************************************/
+static struct no_os_list_desc *actions;
+
+static int32_t irq_action_cmp(void *data1, void *data2)
+{
+	return ((struct irq_action *)data1)->irq_id -
+	       ((struct irq_action *)data2)->irq_id;
+}
+
+static bool initialized[STM32_IRQ_CTRL_NB] =  {false};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/**
+ * @brief Maps pin_nb to NVIC IRQ id
+ * @param pin_nb - pin number to be mapped
+ * @param irq_id - NVIC IRQ id corresponding to given pin_nb
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+static int stm32_pin_get_irq(uint8_t pin_nb, IRQn_Type *irq_id)
+{
+	switch(pin_nb) {
+	case 0:
+		*irq_id = EXTI0_IRQn;
+		break;
+	case 1:
+		*irq_id = EXTI1_IRQn;
+		break;
+	case 2:
+		*irq_id = EXTI2_IRQn;
+		break;
+	case 3:
+		*irq_id = EXTI3_IRQn;
+		break;
+	case 4:
+		*irq_id = EXTI4_IRQn;
+		break;
+	case 5:
+	case 6:
+	case 7:
+	case 8:
+	case 9:
+		*irq_id = EXTI9_5_IRQn;
+		break;
+	case 10:
+	case 11:
+	case 12:
+	case 13:
+	case 14:
+	case 15:
+		*irq_id = EXTI15_10_IRQn;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief EXTI GPIO Interrupt handler callback
+ * @param pin pin number on which the interrupt occurred (GPIO_PIN_pin)
+ */
+void HAL_GPIO_EXTI_Callback(uint16_t pin)
+{
+	int ret;
+	struct irq_action *action;
+	struct irq_action key;
+
+	key.irq_id = no_os_find_first_set_bit(pin);
+
+	if (EXTI->PR & pin) {
+		/* Clear pending bit */
+		EXTI->PR = pin;
+
+		ret = no_os_list_read_find(actions, (void **)&action, &key);
+		if (ret)
+			return;
+
+		if (action->callback)
+			action->callback(action->ctx);
+	}
+}
+
+/**
+ * @brief Initialize the GPIO interrupt controller
+ * @param desc - Pointer where the configured instance is stored
+ * @param param - Configuration information for the instance
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+static int32_t stm32_gpio_irq_ctrl_init(struct no_os_irq_ctrl_desc **desc,
+					const struct no_os_irq_init_param *param)
+{
+	static struct no_os_irq_ctrl_desc *gpio_irq_desc_arr[STM32_IRQ_CTRL_NB];
+
+	int ret;
+	struct no_os_irq_ctrl_desc *gpio_irq_desc;
+	struct stm32_gpio_irq_desc *sdesc;
+	struct stm32_gpio_irq_init_param *gpio_irq_ip;
+	uint8_t gpio_irq_ctrl_id;
+
+	if (!param || !param->extra)
+		return -EINVAL;
+
+	gpio_irq_ip =  param->extra;
+
+	/* Check port number */
+	if (!IS_EXTI_GPIO_PORT(gpio_irq_ip->port_nb))
+		return -EINVAL;
+
+	/* Check pin number */
+	if (!IS_EXTI_GPIO_PIN(param->irq_ctrl_id))
+		return -EINVAL;
+
+	/* Obtain gpio irq controller id */
+	if (param->irq_ctrl_id <= 4)
```

Comment:
same comment here, i think this would work well for the families mentioned above, but this mapping is not true for other families

i think there are families with just 3 interrupt events instead of 6 and they map them differently:
- L0 and possibly F0 for example have EXTI0_1_IRQHandler, EXTI2_3_IRQHandler, EXTI4_15_IRQHandler

Reply:
I have made some new changes which no longer use specifically EXTI_n_IRQn in the driver. please take a look and let me know your thoughts!